src/green/builder.rs

```diff
@@ -159,38 +159,40 @@ impl Builder {
 }
 
 impl Builder {
-    fn turn_node_gc(&mut self) -> bool {
+    fn collect_root_nodes(&mut self) -> Vec<Arc<Node>> {
         // NB: `drain_filter` is `retain` but with an iterator of the removed elements.
         // i.e.: elements where the predicate is FALSE are removed and iterated over.
-        self.nodes.drain_filter(|ThinEqNode(node), ()| Arc::strong_count(node) > 1).any(|_| true)
+        self.nodes
+            .drain_filter(|ThinEqNode(node), ()| Arc::strong_count(node) > 1)
+            .map(|(ThinEqNode(node), _)| node)
+            .collect()
     }
 
-    fn turn_token_gc(&mut self) -> bool {
-        self.tokens.drain_filter(|token, ()| Arc::strong_count(token) > 1).any(|_| true)
-    }
-
-    /// Collect cached nodes that are no longer live outside the cache.
-    ///
-    /// This is a single turn of the GC, and may not GC all potentially unused
-    /// nodes in the cache. To run this to a fixpoint, use [`Builder::gc`].
-    pub fn turn_gc(&mut self) -> bool {
-        let removed_nodes = self.turn_node_gc();
-        let removed_tokens = self.turn_token_gc();
-        removed_nodes || removed_tokens
+    fn collect_tokens(&mut self) {
+        self.tokens.retain(|token, ()| Arc::strong_count(token) > 1)
     }
 
     /// Collect all cached nodes that are no longer live outside the cache.
-    ///
-    /// This is slightly more efficient than just running [`Builder::turn_gc`]
-    /// to a fixpoint, as it knows more about the cache structure and can avoid
-    /// re-GCing definitely clean sections.
     pub fn gc(&mut self) {
-        // Nodes can keep other elements live, so GC them to a fixpoint
-        while self.turn_node_gc() {
-            continue;
+        let mut to_drop = self.collect_root_nodes();
+        let Builder { hasher, nodes, .. } = self;
+
+        while let Some(node) = to_drop.pop() {
+            for child in node.children() {
+                if let Some(node) = child.into_node() {
+                    to_drop.push(ArcBorrow::upgrade(node));
+                }
+            }
```

Comment:
Should this not be inside the strong count check? Otherwise you check the entire tree of nodes that need to stay.

Reply:
You are correct; changed in #56 

I was incorrectly thinking that I had to always recurse into children to handle the case where a child node showed up more than once in the tree. This is just handled by the fact that the node will show up again in the processing stack when the second instance is reached, however.